Mathlib/Data/List/OfFn.lean

```diff
@@ -181,6 +181,63 @@ theorem pairwise_ofFn {R : α → α → Prop} {n} {f : Fin n → α} :
     (Fin.rightInverse_cast (length_ofFn f)).surjective.forall, Fin.forall_iff, Fin.cast_mk,
     Fin.mk_lt_mk, forall_comm (α := (_ : Prop)) (β := ℕ)]
 
+lemma head_ofFn {n} (hn : n ≠ 0) (f : Fin n → α) :
+    (ofFn f).head (mt ofFn_eq_nil_iff.1 hn) = f ⟨0, Nat.pos_of_ne_zero hn⟩ := by
+  rw [← getElem_zero (length_ofFn _ ▸ Nat.pos_of_ne_zero hn), List.getElem_ofFn]
+
+lemma getLast_ofFn {n} (hn : n ≠ 0) (f : Fin n → α) :
+    (ofFn f).getLast (mt ofFn_eq_nil_iff.1 hn) = f ⟨n - 1, Nat.sub_one_lt hn⟩ := by
+  simp_rw [getLast_eq_getElem, length_ofFn, List.getElem_ofFn]
+
+lemma ofFn_cons {n} (a : α) (f : Fin n → α) : ofFn (Fin.cons a f) = a :: ofFn f := by
+  rw [ofFn_succ]
+  rfl
+
+lemma find?_ofFn_eq_some {n} {f : Fin n → α} {p : α → Bool} {b : α} :
+    (ofFn f).find? p = some b ↔ p b = true ∧ ∃ i, f i = b ∧ ∀ j < i, ¬(p (f j) = true) := by
```

Comment:
This seems like it should follow from [`List.ofFn_eq_pmap`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/List/FinRange.html#List.ofFn_eq_pmap) and [`List.find?_range_eq_some`](https://leanprover-community.github.io/mathlib4_docs/Init/Data/List/Nat/Range.html#List.find?_range_eq_some).

Reply:
`ofFn_eq_pmap` doesn't leave things in a good form for applying `find?_range_eq_some` (either on its own or with `find?_pmap` applied in between). If there are other API lemmas (existing or new) that can be applied to achieve the translation to `find?` on `range`, it's not clear to me what they would be.

Reply:
```
theorem find?_eq_some' {xs : List α} {p : α → Bool} {b : α} : 
    xs.find? p = some b ↔ p b ∧ ∃ i h, xs[i] = b ∧ ∀ j : Nat, (hj : j < i) → !p (xs[j]'(sorry))
```    
would be a better intermediate step.

Reply:
```
theorem find?_eq_some_iff_getElem {xs : List α} {p : α → Bool} {b : α} :
    xs.find? p = some b ↔ p b ∧ ∃ i h, xs[i] = b ∧ ∀ j : Nat, (hj : j < i) → !p xs[j] := by
  rw [find?_eq_some]
  simp only [Bool.not_eq_eq_eq_not, Bool.not_true, exists_and_right, and_congr_right_iff]
  intro w
  constructor
  · rintro ⟨as, ⟨bs, rfl⟩, h⟩
    refine ⟨as.length, ⟨?_, ?_, ?_⟩⟩
    · simp only [length_append, length_cons]
      refine Nat.lt_add_of_pos_right (zero_lt_succ bs.length)
    · rw [getElem_append_right (Nat.le_refl as.length)]
      simp
    · intro j h'
      rw [getElem_append_left h']
      exact h _ (getElem_mem h')
  · rintro ⟨i, h, rfl, h'⟩
    refine ⟨xs.take i, ⟨xs.drop (i+1), ?_⟩, ?_⟩
    · rw [getElem_cons_drop, take_append_drop]
    · intro a m
      rw [mem_take_iff_getElem] at m
      obtain ⟨j, h, rfl⟩ := m
      apply h'
      omega
```

I'll put this in Lean, but if you want to put it in Mathlib in the meantime that's fine.

(I might also rename the existing `find?_eq_some` as ` find?_eq_some_iff_append` to more clearly distinguish them.)

Reply:
I've copied that lemma into `Mathlib.Data.List.OfFn` (as a temporary location since there isn't an obvious better one imported by that file) and refactored `find?_ofFn_eq_some` in terms of it.